File: packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebase/auth/FlutterFirebaseTotpMultiFactor.java
/*
 * Copyright 2023, the Chromium project authors.  Please see the AUTHORS file
 * for details. All rights reserved. Use of this source code is governed by a
 * BSD-style license that can be found in the LICENSE file.
 */

package io.flutter.plugins.firebase.auth;

import androidx.annotation.NonNull;
import com.google.firebase.auth.MultiFactorSession;
import com.google.firebase.auth.TotpMultiFactorAssertion;
import com.google.firebase.auth.TotpMultiFactorGenerator;
import com.google.firebase.auth.TotpSecret;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

public class FlutterFirebaseTotpMultiFactor
    implements GeneratedAndroidFirebaseAuth.MultiFactorTotpHostApi {

  // Map an app id to a map of user id to a TotpSecret object.
  static final Map<String, TotpSecret> multiFactorSecret = new HashMap<>();

  @Override
  public void generateSecret(
      @NonNull String sessionId,
      @NonNull
          GeneratedAndroidFirebaseAuth.Result<GeneratedAndroidFirebaseAuth.PigeonTotpSecret>
              result) {
    MultiFactorSession multiFactorSession =
        FlutterFirebaseMultiFactor.multiFactorSessionMap.get(sessionId);

    assert multiFactorSession != null;
    TotpMultiFactorGenerator.generateSecret(multiFactorSession)
        .addOnCompleteListener(
            task -> {
              if (task.isSuccessful()) {
                TotpSecret secret = task.getResult();
                multiFactorSecret.put(secret.getSharedSecretKey(), secret);
                result.success(
                    new GeneratedAndroidFirebaseAuth.PigeonTotpSecret.Builder()
                        .setCodeIntervalSeconds((long) secret.getCodeIntervalSeconds())
                        .setCodeLength((long) secret.getCodeLength())
                        .setSecretKey(secret.getSharedSecretKey())
                        .setHashingAlgorithm(secret.getHashAlgorithm())
                        .setEnrollmentCompletionDeadline(secret.getEnrollmentCompletionDeadline())
                        .build());
              } else {
                result.error(
                    FlutterFirebaseAuthPluginException.parserExceptionToFlutter(
                        task.getException()));
              }
            });
  }

  @Override
  public void getAssertionForEnrollment(
      @NonNull String secretKey,
      @NonNull String oneTimePassword,
      @NonNull GeneratedAndroidFirebaseAuth.Result<String> result) {
    final TotpSecret secret = multiFactorSecret.get(secretKey);

    assert secret != null;
    TotpMultiFactorAssertion assertion =
        TotpMultiFactorGenerator.getAssertionForEnrollment(secret, oneTimePassword);
    String assertionId = UUID.randomUUID().toString();
    FlutterFirebaseMultiFactor.multiFactorAssertionMap.put(assertionId, assertion);
    result.success(assertionId);
  }

  @Override
  public void getAssertionForSignIn(
      @NonNull String enrollmentId,
      @NonNull String oneTimePassword,
      @NonNull GeneratedAndroidFirebaseAuth.Result<String> result) {
    TotpMultiFactorAssertion assertion =
        TotpMultiFactorGenerator.getAssertionForSignIn(enrollmentId, oneTimePassword);
    String assertionId = UUID.randomUUID().toString();
    FlutterFirebaseMultiFactor.multiFactorAssertionMap.put(assertionId, assertion);
    result.success(assertionId);
  }
}

File: packages/firebase_auth/firebase_auth/android/src/main/kotlin/io/flutter/plugins/firebase/auth/FlutterFirebaseTotpMultiFactor.kt
/*
 * Copyright 2023, the Chromium project authors.  Please see the AUTHORS file
 * for details. All rights reserved. Use of this source code is governed by a
 * BSD-style license that can be found in the LICENSE file.
 */
package io.flutter.plugins.firebase.auth

import com.google.firebase.auth.TotpMultiFactorGenerator
import com.google.firebase.auth.TotpSecret
import java.util.UUID

class FlutterFirebaseTotpMultiFactor : MultiFactorTotpHostApi {
  override fun generateSecret(sessionId: String, callback: (Result<InternalTotpSecret>) -> Unit) {
    val multiFactorSession = FlutterFirebaseMultiFactor.multiFactorSessionMap[sessionId]
    checkNotNull(multiFactorSession)
    TotpMultiFactorGenerator.generateSecret(multiFactorSession).addOnCompleteListener { task ->
      if (task.isSuccessful) {
        val secret = task.result
        multiFactorSecret[secret.sharedSecretKey] = secret
        callback(
            Result.success(
                InternalTotpSecret(
                    codeIntervalSeconds = secret.codeIntervalSeconds.toLong(),
                    codeLength = secret.codeLength.toLong(),
                    secretKey = secret.sharedSecretKey,
                    hashingAlgorithm = secret.hashAlgorithm,
                    enrollmentCompletionDeadline = secret.enrollmentCompletionDeadline)))
      } else {
        callback(
            Result.failure(
                FlutterFirebaseAuthPluginException.parserExceptionToFlutter(task.exception)))
      }
    }
  }

  override fun getAssertionForEnrollment(
      secretKey: String,
      oneTimePassword: String,
      callback: (Result<String>) -> Unit
  ) {
    val secret = multiFactorSecret[secretKey]
    checkNotNull(secret)
    val assertion = TotpMultiFactorGenerator.getAssertionForEnrollment(secret, oneTimePassword)
    val assertionId = UUID.randomUUID().toString()
    FlutterFirebaseMultiFactor.multiFactorAssertionMap[assertionId] = assertion
    callback(Result.success(assertionId))
  }

  override fun getAssertionForSignIn(
      enrollmentId: String,
      oneTimePassword: String,
      callback: (Result<String>) -> Unit
  ) {
    val assertion = TotpMultiFactorGenerator.getAssertionForSignIn(enrollmentId, oneTimePassword)
    val assertionId = UUID.randomUUID().toString()
    FlutterFirebaseMultiFactor.multiFactorAssertionMap[assertionId] = assertion
    callback(Result.success(assertionId))
  }

  companion object {
    val multiFactorSecret: MutableMap<String, TotpSecret> = HashMap()
  }
}
